pkg/events/queue/queue_mem_list.go

```diff
@@ -0,0 +1,109 @@
+package queue
+
+import (
+	"container/list"
+	"github.com/aquasecurity/tracee/pkg/external"
+	"sync"
+)
+
+type EventQueueMem struct {
+	amountOfEvents     int // number of cached events
+	mutex              sync.Mutex
+	cache              *list.List
+	EventsCacheMemSize int
+}
+
+func (q *EventQueueMem) String() string {
+	return "In-Memory Event Queue"
+}
+
+func (q *EventQueueMem) Setup() error {
+	q.amountOfEvents = q.getQueueSizeInEvents()
+	q.cache = new(list.List)
+
+	return nil
+}
+
+// available returns when queue is available for enqueuing events
+func (q *EventQueueMem) available() {
+	for {
```

Comment:
I think what we are trying to achieve here can be done with condition variables. The variable state is changed by `Dequeue` function (once `q.cache.Len() < q.amountOfEvents`).
While Enqueue is waiting on the condition that `q.cache.Len() < q.amountOfEvents`

Reply:
I'm calling this function outside lock scope from Enqueue so I can lock and unlock while looping here (to allow dequeueing, while still checking the length).

In between one check and another there might be more than 1 dequeues, because of mutex fights, thus the loop. If I do 1 dequeue per 1 enqueue, I'll make enqueue/dequeueing synchronous whenever the buffer is full.. its a different approach then this one.

Now, if there is a better implementation inside Mutex class, like with lighter locking primitives, watching a variable state/change and waking up a callback, or something like a read-copy-update approach, then I can use here I suppose. 

One clear improvement for this would be to use RO locks for checking the length, as you said. Or we could yield sometime so the Dequeue would have more time to do chunks of dequeuing (this would reduce CPU consumption, but would impact throughput).

Quick note: when looping here, I'm still holding the event, instead of returning and loosing the event.

Reply:
NM, I just read what you wrote correctly. I believe you are suggesting having a standalone variable that won't compete with the locking, I think that is good, will do it.